In [65]:
import pandas as pd
import sklearn as sk

2 .Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

Our model is trying to correctly identify dogs.

In the context of this problem, what is a false positive?

-- We predicted the animal was a dog but it was a act. There were 13 false positives.

In the context of this problem, what is a false negative?

-- We predicted the animal was a cat but it was actualy a dog. There were 7 false negatives in this data set.



#### Evaluating the model

Total number of observations == 100

True Positive == 46  The model predected dog and it was a dog.

False Positive == 13  The model predicted dog but it was actually a cat.

True Negative == 34  The model predicted cat and it was actually a cat.

False Negative == 7  The model predicted cat and it was actually a dog.

Accuracy == 46 + 34 / 100 == 80/100 == 4/5 == 0.8 == (TP + TN) / (TP + TN + FP + FN)

Precision == 46 / 46 + 13 == 46 / 59 == .78 == TP / (TP + FP)

Recall == 46 / 46 + 7 == 46 / 53 == .87 == TP / (TP + FN)

#### How would you describe this model?

This model is very good at capturing the positive rates with a recall rate of 87%.

This model is very accurate with an acccuracy of 80%.

## 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

In [12]:
c3 = pd.read_csv('/Users/edwige/downloads/c3.csv')
c3.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [21]:
c3.actual.value_counts()

actual
No Defect    184
Defect        16
Name: count, dtype: int64

In [25]:
pd.crosstab(c3.actual, c3.model1)

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [22]:
pd.crosstab(c3.actual, c3.model2)

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [23]:
pd.crosstab(c3.actual, c3.model3)

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

### An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

-- Model3 woudl be the best model to identiy defects as it correctly identified 13 out of 16 defects. With a recall evaluation of .81

-- This would be a good test of recall

 Recall Tests
 TP / (TP + FN)
 
 -- M1 == 8 / 8 + 8 == .5
 
 -- M2 == 9 / 9 + 7 == .56
 
 -- M3 == 13 / 13 + 3 == .81

### Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

-- This would be a good test of precision

-- Model1 would be the best fit for this as it has the highest accuracy of .8

Precicion tests
TP / (TP + FP)

-- M1 == 8 / 8 + 2 == .8

-- M2 == 9 / 9 + 81 == .1

-- M3 == 13 / 13 + 86 == .13

4. Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [32]:
paws = pd.read_csv('/Users/edwige/downloads/gives_you_paws.csv')
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [37]:
paws.actual.value_counts()

actual
dog    3254
cat    1746
Name: count, dtype: int64

In [38]:
paws['baseline'] = 'dog'
paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [33]:
pd.crosstab(paws.actual,paws.model1)

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [34]:
pd.crosstab(paws.actual,paws.model2)

model2   cat   dog
actual            
cat     1555   191
dog     1657  1597

In [35]:
pd.crosstab(paws.actual,paws.model3)

model3   cat   dog
actual            
cat      893   853
dog     1599  1655

In [36]:
pd.crosstab(paws.actual,paws.model4)

model4  cat   dog
actual           
cat     603  1143
dog     144  3110

In [39]:
pd.crosstab(paws.actual, paws.baseline)

baseline   dog
actual        
cat       1746
dog       3254

#### a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

-- Both model1(.81) and model4(.74) outperform the baseline(.65) for accuracy.

Accuracy
(TP + TN) / (TP + TN + FP + FN)

-- M1 (2614 + 1423) / (5000) == .81

-- M2 (1597 + 1555) / (5000) == .63

-- M3 (1655 + 893) / (5000) == .51

-- M4 (3110 + 603) / (5000) == .74

-- Base (3254) / (5000) == .65

#### b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?


-- This would require a test of precision

-- Both Model1 and Model2 would be a good fit for selecting dog pictures with precision.

Precision
TP / (TP + FP)

-- M1 2614 / (2614 + 323) == 0.890

-- M2 1597 / (1597 + 191) == 0.893

-- M3 1655 / (1655 + 853) == 0.660

-- M4 3110 / (3110 + 1143) == 0.731

-- Base 3259 / 5000 == 0.652


#### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

-- This would require a test of precision

-- Model4 would be a good fit for selecting cat pictures with precision of .807.

Precision
TP / (TP + FP)
-- M1 1423 / (1423 + 640) == .690

-- M2 1555 / (1555 + 1657) == .484

-- M3 893 / (893 + 1599) == .358

-- M4 603 / (603 + 144) == .807

-- Base 0 

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

#### Accuracy score